diff -Naur kodi-15.2-02e7013/configure.ac kodi.work/configure.ac
--- kodi-15.2-02e7013/configure.ac	2016-03-18 11:36:06.171253305 +0100
+++ kodi.work/configure.ac	2016-03-18 11:35:25.083253784 +0100
@@ -1952,6 +1952,10 @@
         CFLAGS="$CFLAGS -DEGL_API_FB"
       fi
       ;;
+    *hisilicon*)
+      AC_CHECK_LIB([player], main, LIBS="$LIBS -lplayer -lhi_so -lsubdec", AC_MSG_ERROR($missing_library))
+      AC_CHECK_LIB([hi_sample_common], main, LIBS="$LIBS -lhi_msp -ljpeg6b -lhi_common", AC_MSG_ERROR($missing_library))
+      ;;
     *)
   esac
 done

